layers: Add VK_KHR_dynamic_rendering

AaronHaganAMD opened the pull request:
WIP - Add VU for VK_KHR_dynamic_rendering extension.

ci-tester-lunarg commented:
Author AaronHaganAMD not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author AaronHaganAMD not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author AaronHaganAMD not on autobuild list. Waiting for curator authorization before starting CI build.

jeremyg-lunarg commented on layers/core_validation.h:
```diff
@@ -1081,6 +1081,10 @@ class CoreChecks : public ValidationStateTracker {
+    bool PreCallValidateCmdBeginRenderingKHR(VkCommandBuffer commandBuffer,
+                                             const VkRenderingInfoKHR* pRenderingInfo) const override;
+    bool PreCallValidateRenderingAttachmentInfoKHR(VkCommandBuffer commandBuffer,
+                                                    const VkRenderingAttachmentInfoKHR* pAttachments) const;
```
This needs to be 'const override'

jeremyg-lunarg commented on layers/state_tracker.cpp:
```diff
@@ -2896,6 +2904,28 @@ void ValidationStateTracker::PostCallRecordCmdEndConditionalRenderingEXT(VkComma
+void ValidationStateTracker::RecordCmdBeginRenderingRenderPassState(VkCommandBuffer commandBuffer,
+    const VkRenderingInfoKHR* pRenderingInfo) {
+    auto* cb_state = Get<CMD_BUFFER_STATE>(commandBuffer);
+    // cb_state->useDynamicRendering = true;
```
these commented out lines need to be removed or uncommented, not sure which.

jeremyg-lunarg commented on scripts/known_good.json:
```diff
@@ -18,7 +18,7 @@
       "build_dir": "Vulkan-Headers/build",
       "install_dir": "Vulkan-Headers/build/install",
-      "commit": "v1.2.196"
+      "commit": "v1.2.197"
```
This should go away if you rebase on top of def5cf6be051fd791e9c6bf4a41d4b6617fcbbf8

jeremyg-lunarg commented on layers/state_tracker.cpp:
```diff
@@ -1927,9 +1927,17 @@ bool ValidationStateTracker::PreCallValidateCreateGraphicsPipelines(VkDevice dev
-                                                                          GetPipelineLayoutShared(pCreateInfos[i].layout)));
+        auto begin_rendering_info = LvlFindInChain<VkPipelineRenderingCreateInfoKHR>(pCreateInfos[i].pNext);
+        if (begin_rendering_info) {
+            std::shared_ptr<RENDER_PASS_STATE> renderPassState(new RENDER_PASS_STATE(begin_rendering_info));
```
should be render_pass_state. yes we have lots of inconsistencies but trying to get better :)

jeremyg-lunarg commented on layers/render_pass_state.h:
```diff
@@ -88,6 +88,9 @@ class RENDER_PASS_STATE : public BASE_NODE {
         AttachmentTransition(uint32_t prev_pass_, uint32_t attachment_, VkImageLayout old_layout_, VkImageLayout new_layout_)
             : prev_pass(prev_pass_), attachment(attachment_), old_layout(old_layout_), new_layout(new_layout_) {}
     };
+    bool useDynamicRendering;
```
these struct members should be snake_case names.

jeremyg-lunarg replied:
also, can the non-const members be made const?

ci-tester-lunarg commented:
Author AaronHaganAMD not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author AaronHaganAMD not on autobuild list. Waiting for curator authorization before starting CI build.

jeremyg-lunarg reviewed (commented):
Good to see this PR! Some preliminary comments since it looks like you're still working on it. The 2 big things so far are:

1.  Please rebase against the current tip of master, which is already updated to Vulkan-Headers 1.2.197. This should make the size  of your patch go way down.
2.  Please use snake_case for struct members and local variables. Yes the code is very inconsistent but we're trying to move to using the Google style guide (see:  CONTRIBUTING.md)

ci-tester-lunarg commented:
Author AaronHaganAMD not on autobuild list. Waiting for curator authorization before starting CI build.

sfricke-samsung commented on layers/core_validation.cpp:
```diff
@@ -6367,6 +6386,429 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+                        pRenderingInfo->viewMask, pRenderingInfo->layerCount);
+    }
+
+    if (!(device_extensions.vk_amd_mixed_attachment_samples || device_extensions.vk_nv_framebuffer_mixed_samples)) {
```
for anything with `device_extensions.vk_*` please use the `IsExtEnabled()` wrapper

sfricke-samsung commented on layers/core_validation.cpp:
```diff
@@ -6367,6 +6386,429 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+
+            if (first_sample_count_attachment != image_view->samples) {
+                skip |= LogError(commandBuffer, "VUID-VkRenderingInfoKHR-imageView-06070",
+                                 "Color attachment ref %u has sample count %s, whereas first color attachment ref has "
```
please add `vkCmdBeginRenderingKHR(): ` to start each error message

jeremyg-lunarg commented on layers/pipeline_state.h:
```diff
@@ -190,6 +204,7 @@ class PIPELINE_STATE : public BASE_NODE {
         safe_VkGraphicsPipelineCreateInfo graphics;
         safe_VkComputePipelineCreateInfo compute;
         safe_VkRayTracingPipelineCreateInfoCommon raytracing;
+        safe_VkPipelineRenderingCreateInfoKHR dynamic_rendering;
```
This member isn't used and I don't think it would work to add it here.  This structure is chained onto the graphics pipeline info, so it should be available through graphics.pNext

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 14472.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5448 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 14483.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5449 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5449 failed.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6393,477 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+            if (pRenderingInfo->pDepthAttachment->imageLayout == VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL) {
+                skip |= LogError(
+                    commandBuffer, "VUID-VkRenderingInfoKHR-pDepthAttachment-06092",
+                    "vkCmdBeginRenderingKHR(): image must not have been created with VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL");
```
imageLayout must not be

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6393,477 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+                }
+            }
+
+            if (pRenderingInfo->pDepthAttachment->resolveMode != VK_RESOLVE_MODE_NONE) {
```
combine the 2 resolveMode != NONE if conditionals?

Tobski replied:
Done.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6393,477 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+            if (pRenderingInfo->pStencilAttachment->imageLayout == VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL) {
+                skip |= LogError(
+                    commandBuffer, "VUID-VkRenderingInfoKHR-pStencilAttachment-06094",
+                    "vkCmdBeginRenderingKHR(): image must not have been created with VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL");
```
imageLayout must not be

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6393,477 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+
+    if (pAttachments->imageView != VK_NULL_HANDLE) {
+        auto image_view_state = GetImageViewState(pAttachments->imageView);
+        if (pAttachments->resolveMode != VK_RESOLVE_MODE_NONE && image_view_state->samples != VK_SAMPLE_COUNT_1_BIT) {
```
combine all of the resolveMode != NONE conditionals?

Tobski replied:
Done

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6414,6 +6904,22 @@ bool CoreChecks::PreCallValidateBeginCommandBuffer(VkCommandBuffer commandBuffer
                     }
                 }
+                const auto *renderPass = GetRenderPassState(info->renderPass);
+                if (renderPass && !renderPass->use_dynamic_rendering) {
```
I think this needs to be:
```
       if (info->renderPass != VK_NULL_HANDLE) {
           const auto *render_pass = GetRenderPassState(info->renderPass);
           if (!render_pass) {
                  VUID-06000
            }
       } else {
```

jeremyg-lunarg reviewed (changes requested):
Whew, made it all the way through the changes. Other than the extra formatting changes that got added in the last push, I think this is getting close. Several comments inline about some of the VUID checks and other things.

AaronHaganAMD commented on layers/best_practices_utils.cpp:
```diff
@@ -2337,7 +2337,7 @@ void BestPractices::PreCallRecordCmdClearAttachments(VkCommandBuffer commandBuff
                                                      uint32_t rectCount, const VkClearRect* pRects) {
     auto* cmd_state = GetCBState(commandBuffer);
-    RENDER_PASS_STATE* rp_state = cmd_state->activeRenderPass.get();
+    RENDER_PASS_STATE* rp_state = cmd_state->active_render_pass.get();
```
I can leave the pre-existing members alone for now..

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 14613.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5451 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 14623.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5452 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5452 failed.

Tobski commented on layers/render_pass_state.h:
```diff
@@ -88,6 +88,9 @@ class RENDER_PASS_STATE : public BASE_NODE {
             : prev_pass(prev_pass_), attachment(attachment_), old_layout(old_layout_), new_layout(new_layout_) {}
     };
+    const bool use_dynamic_rendering;
+    safe_VkRenderingInfoKHR dynamic_rendering_info;
```
It's updated in

```c++
RENDER_PASS_STATE::RENDER_PASS_STATE(VkRenderingInfoKHR const *pRenderingInfo);
```

which is called by 

```c++
void CMD_BUFFER_STATE::BeginRendering(...)
```

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5453 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 14816.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5454 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5454 failed.

jeremyg-lunarg commented:
@AaronHaganAMD and @Tobski do you have an estimate when you'll be done adding functionality to this PR? The deadline to get it stabilized, cleaned up and merged for the next SDK is rapidly approaching.  It is currently crashing the unit tests on all targets in VkLayerTest.SecondaryCommandBufferNullRenderpass.  If you need details about the crashes or cannot reproduce them locally please let me know.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 14906.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5458 running.

Tobski commented:
> @AaronHaganAMD and @Tobski do you have an estimate when you'll be done adding functionality to this PR? The deadline to get it stabilized, cleaned up and merged for the next SDK is rapidly approaching. It is currently crashing the unit tests on all targets in VkLayerTest.SecondaryCommandBufferNullRenderpass. If you need details about the crashes or cannot reproduce them locally please let me know.

We're working more or less full time on this right now, we'll be done before the end of the week (I believe your deadline is Friday?)

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5458 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 14950.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 14959.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5460 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 14983.

jeremyg-lunarg commented:
> We're working more or less full time on this right now, we'll be done before the end of the week (I believe your deadline is Friday?)

Ok. I do need some time to review and to run additional tests beyond our automated CI on the final version, which will take several hours.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5462 running.

KarenGhavam-lunarG commented:
> We're working more or less full time on this right now, we'll be done before the end of the week (I believe your deadline is Friday?)

We are suddenly being asked for lots of code reviews on a compressed schedule due to the quantity of changes coming in late. To give us some time to do the reviews and for you to be able to react and fix per the reviews, you should not be thinking that Friday is your deadline. More like EOB today.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5462 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 15038.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5465 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 15094.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 15100.

AaronHaganAMD commented:
VkLayerTest.SecondaryCommandBufferNullRenderpass should be passing now..

jeremyg-lunarg commented:
> VkLayerTest.SecondaryCommandBufferNullRenderpass should be passing now..

Great! Are you all done making functional changes or is there still more to come?

AaronHaganAMD commented:
> > VkLayerTest.SecondaryCommandBufferNullRenderpass should be passing now..
> 
> Great! Are you all done making functional changes or is there still more to come?

I think this PR is in good shape.. But there are still a lot of VUs that still need added. Should we get this PR merged and add more VU's in a new a PR ?

jeremyg-lunarg commented:
> 
> I think this PR is in good shape.. But there are still a lot of VUs that still need added. Should we get this PR merged and add more VU's in a new a PR ?

Definitely needs to be a separate PR for the additional VUs at this point. I'll start reviewing this one again.

Tobski commented:
Just to be clear, when I said "end of Friday" I was including review time - I didn't foresee us having substantial changes beyond EOB today. Tomorrow morning I'll take a look at @jeremyg-lunarg's review comments, and hopefully when you all come online tomorrow we'll be in good shape to submit - everything from then til Friday is wiggle room.

jeremyg-lunarg commented on layers/buffer_validation.cpp:
```diff
@@ -3480,37 +3479,89 @@ void CoreChecks::PreCallRecordCmdClearAttachments(VkCommandBuffer commandBuffer,
+                        image_index = clear_desc->colorAttachment;
+                    }
+                    else if (clear_desc->aspectMask & (VK_IMAGE_ASPECT_DEPTH_BIT)) {
+                        image_index = colorAttachmentCount * 2;
```
since this offset math has to stay in sync with the code in CMD_BUFFER_STATE::BeginRendering(), would it be possible to have a method, function or something to return the correct index, which gets used in both places?

Tobski replied:
Yea that's probably not the worst idea. I think there's scope  for a lot of refactoring in this area which I think is something we probably want to discuss after this lands.

Tobski replied:
Done, I've added functions to cmd_buffer_state for this.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -1363,16 +1376,22 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
+    safe_VkSubpassDescription2 *subpass_desc = nullptr;
+
+    if (pPipeline->rp_state.get() == nullptr) {
+        // Validate dynamic rendering
```
seems like this should be checking rp_state->use_dynamic_rendering since there will be a stub rp_state  here.  Are there VUIDs that you're going to add here in a separate PR? If so maybe make this a TODO

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6394,475 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+                    if (!(image_state->createInfo.extent.height >= offset_y + height)) {
+                        skip |= LogError(
+                            commandBuffer, "VUID-VkRenderingInfoKHR-pNext-06084",
+                            "vkCmdBeginRenderingKHR(): width of the pColorAttachments[%u].imageView: %u must be greater than"
```
this error is for a height check but the message says width.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6394,475 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+                }
+                if (!(image_state->createInfo.extent.height >= offset_y + height)) {
+                    skip |= LogError(commandBuffer, "VUID-VkRenderingInfoKHR-pNext-06084",
+                                     "vkCmdBeginRenderingKHR(): width of the pDepthAttachment->imageView: %u must be greater than"
```
also should be height.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6394,475 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+                }
+                if (!(image_state->createInfo.extent.height >= offset_y + height)) {
+                    skip |= LogError(commandBuffer, "VUID-VkRenderingInfoKHR-pNext-06084",
+                                     "vkCmdBeginRenderingKHR(): width of the pStencilAttachment->imageView: %u must be greater than"
```
height

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6394,475 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+                (pRenderingInfo->pDepthAttachment->resolveMode != VK_RESOLVE_MODE_NONE) &&
+                (pRenderingInfo->pStencilAttachment->resolveMode != VK_RESOLVE_MODE_NONE) &&
+                (pRenderingInfo->pStencilAttachment->resolveMode != pRenderingInfo->pDepthAttachment->resolveMode)) {
+                skip |= LogError(device, "VUID-VkRenderingInfoKHR-pDepthAttachment-06105",
```
06105 and 06104 seem to contradict each other, if resolveMode is NONE for depth or stencil, you'd fail 06104 but pass 06105. Is this a spec problem?  Seems like it would be less confusing to only check 06105 here.

Tobski replied:
I'm not sure why you think this is a contradiction - these are two separate limit values being tested (albeit with very similar names). We could perhaps change it so that the less restrictive condition is only checked if the more restrictive condition passes?

Tobski replied:
>So is it valid for either resolveMode to be NONE if the other is not? If yes, then 06104 is wrong, if no, then 06105 is redundant. 

Both? Either? Depends on the implementation. Again - separate limits!

6104 triggers if and only if (phys_dev_props_core12.independentResolveNone == VK_FALSE).

6105 triggers if and only if (phys_dev_props_core12.independentResolve == VK_FALSE). Note that this is a _different limit_. In addition, you can't support the former limit if this limit is not supported.

The truth table is thus much bigger... I can type the whole thing out if I absolutely must but I'm not super keen on doing so 😅

jeremyg-lunarg replied:
Ah, I missed the independentResolveNone vs independentResolve. Thanks!

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5470 running.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6394,475 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+            if (pRenderingInfo->pColorAttachments[j].imageLayout == VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL ||
+                pRenderingInfo->pColorAttachments[j].imageLayout == VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL) {
+                skip |= LogError(commandBuffer, "VUID-VkRenderingInfoKHR-colorAttachmentCount-06090",
+                                 "vkCmdBeginRenderingKHR(): imageLayout must not have be "
```
typo: must not have be

Tobski replied:
Fixed.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6394,475 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+        auto resolve_view_state = GetImageViewState(pAttachments->resolveImageView);
+        if (pAttachments->resolveMode != VK_RESOLVE_MODE_NONE && resolve_view_state->samples != VK_SAMPLE_COUNT_1_BIT) {
+            skip |= LogError(commandBuffer, "VUID-VkRenderingAttachmentInfoKHR-imageView-06133",
+                             "resolveImageView sample count must be must not have a VK_SAMPLE_COUNT_1_BIT for Resolve Mode %s",
```
typo: "must be must not have" from the VUID it looks like this should say "must be"

Tobski replied:
Fixed.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -851,48 +851,52 @@ bool CoreChecks::ValidatePipelineDrawtimeState(const LAST_BOUND_STATE &state, co
-                        attachment < pPipeline->create_info.graphics.pColorBlendState->attachmentCount) {
-                        if ((imageview_state->format_features & VK_FORMAT_FEATURE_COLOR_ATTACHMENT_BLEND_BIT) == 0 &&
+            if (pCB->activeRenderPass->use_dynamic_rendering) {
+                const auto dynamic_rendering_info = pCB->activeRenderPass->dynamic_rendering_info;
```
Are there VUIDs you're going to add here in a separate PR? If so maybe make this a TODO

Tobski replied:
I've added a TODO - hopefully that's what you expected?

jeremyg-lunarg commented on layers/state_tracker.cpp:
```diff
@@ -1814,9 +1812,16 @@ bool ValidationStateTracker::PreCallValidateCreateGraphicsPipelines(VkDevice dev
+            std::shared_ptr<RENDER_PASS_STATE> render_pass_state(new RENDER_PASS_STATE(begin_rendering_info));
+            cgpl_state->pipe_state.push_back(std::make_shared<PIPELINE_STATE>(this, &pCreateInfos[i], render_pass_state,
+                                                                              GetPipelineLayoutShared(pCreateInfos[i].layout)));
+        } else {
```
There's going to be a merge conflict from PR #3523 here.  To resolve, I think you should have something like:
```
     if (begin_rendering_info) {
    } else if (pCreateInfos[i].renderPass != VK_NULL_HANDLE) {
    } else {
       insert null shared_ptr.
    }

```

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6394,475 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+                      pRenderingInfo->renderArea.offset.y + pRenderingInfo->renderArea.extent.height)) {
+                    skip |=
+                        LogError(commandBuffer, "VUID-VkRenderingInfoKHR-pNext-06080",
+                                 "vkCmdBeginRenderingKHR(): width of the pColorAttachments[%u].imageView: %u must be greater than"
```
This should say height, not width.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6394,475 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+            if (!(image_state->createInfo.extent.height >=
+                  pRenderingInfo->renderArea.offset.y + pRenderingInfo->renderArea.extent.height)) {
+                skip |= LogError(commandBuffer, "VUID-VkRenderingInfoKHR-pNext-06080",
+                                 "vkCmdBeginRenderingKHR(): width of the pDepthAttachment->imageView: %u must be greater than"
```
height, not width.

jeremyg-lunarg commented on layers/core_validation.cpp:
```diff
@@ -6373,6 +6394,475 @@ bool CoreChecks::PreCallValidateUpdateDescriptorSets(VkDevice device, uint32_t d
+            if (!(image_state->createInfo.extent.height >=
+                  pRenderingInfo->renderArea.offset.y + pRenderingInfo->renderArea.extent.height)) {
+                skip |= LogError(commandBuffer, "VUID-VkRenderingInfoKHR-pNext-06080",
+                                 "vkCmdBeginRenderingKHR(): width of the pStencilAttachment->imageView: %u must be greater than"
```
height, not width.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5470 passed.

jeremyg-lunarg reviewed (commented):
I'm still waiting on results from extra testing to make sure we're not regressing any of the application traces in our test suite. But I think this is pretty close. There's probably some refactoring that can be done with the existing code to make this all fit together better, but that's going to happen after the fact.

There will be a change you need to make as a result of  PR #3523,  in CoreChecks::PreCallValidateCreateGraphicsPipelines() I added a check for null renderpass handles, you'll want to remove the UNASSIGNED-VkGraphicsPipelineCreateInfo-dynamicRendering-not-supported case but keep the VUID-VkGraphicsPipelineCreateInfo-dynamicRendering-06052 case. There's another conflict in the ValidationStateTracker version of this method that I noted inline. This change was necessary so that we have somewhat OK behavior if someone triggers 06052 or if we have to ship the SDK without this change for some late breaking reason.

jeremyg-lunarg commented:
Also related to VUID-VkGraphicsPipelineCreateInfo-dynamicRendering-06052, is there a VUID for the case where the renderPass handle is null, but there is not a VkPipelineRenderingCreateInfoKHR structure in the pNext chain?  I cannot find one yet, but this case could cause a crash in the layer if not handled correctly.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 15252.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5472 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5472 passed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 15411.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5473 running.

AaronHaganAMD commented:
> I'm still waiting on results from extra testing to make sure we're not regressing any of the application traces in our test suite. But I think this is pretty close. There's probably some refactoring that can be done with the existing code to make this all fit together better, but that's going to happen after the fact.
> 
> There will be a change you need to make as a result of PR #3523, in CoreChecks::PreCallValidateCreateGraphicsPipelines() I added a check for null renderpass handles, you'll want to remove the UNASSIGNED-VkGraphicsPipelineCreateInfo-dynamicRendering-not-supported case but keep the VUID-VkGraphicsPipelineCreateInfo-dynamicRendering-06052 case. There's another conflict in the ValidationStateTracker version of this method that I noted inline. This change was necessary so that we have somewhat OK behavior if someone triggers 06052 or if we have to ship the SDK without this change for some late breaking reason.

Removed VUID-VkGraphicsPipelineCreateInfo-dynamicRendering-06052  and handled the inline comment in regards to creating the pipeline.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5473 failed.

Tobski commented:
> Also related to VUID-VkGraphicsPipelineCreateInfo-dynamicRendering-06052, is there a VUID for the case where the renderPass handle is null, but there is not a VkPipelineRenderingCreateInfoKHR structure in the pNext chain? I cannot find one yet, but this case could cause a crash in the layer if not handled correctly.

There's no VU for it because default values are assumed for VkPipelineRenderingCreateInfoKHR if it's not present. I'll take a look at making this not crash...

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 15481.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5474 running.

Tobski commented:
> There's no VU for it because default values are assumed for VkPipelineRenderingCreateInfoKHR if it's not present. I'll take a look at making this not crash...

I've added code to handle this and I'll add a test for it shortly.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5474 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 15503.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5475 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5475 passed.

Tobski commented:
@jeremyg-lunarg finally looks like we've got CI passing so the question now is what additional tests do you think we need to get in? I'll do the one I said above, what else do you think we need to get this in?

I'm inclined to hold off tests for the new VUs that were added until after we merge this, because that's a big effort, and running against CTS at least proved they don't generate false positives.

I'm similarly inclined to work on doing any further refactoring work after we merge this, probably as a first step before we add more VU checks.

jeremyg-lunarg commented on layers/state_tracker.cpp:
```diff
@@ -1820,7 +1818,12 @@ bool ValidationStateTracker::PreCallValidateCreateGraphicsPipelines(VkDevice dev
+            cgpl_state->pipe_state.push_back(std::make_shared<PIPELINE_STATE>(this, &pCreateInfos[i],
+                                                                              std::make_shared<RENDER_PASS_STATE>(dynamic_rendering),
+                                                                              GetPipelineLayoutShared(pCreateInfos[i].layout)));
+        } else if (dynamic_rendering || pCreateInfos[i].renderPass != VK_NULL_HANDLE) {
```
This isn't correct and I think it is causing a crash in a test case I wrote yesterday evening.  Probably it needs to be if (!dynamic_rendering || ...) See https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/3525  VkLayerTest.CreateGraphicsPipelineDynamicRenderingNoInfo

jeremyg-lunarg replied:
actually, this seems to work better for me:
```
        if (pCreateInfos[i].renderPass != VK_NULL_HANDLE) {
              cgpl_state->pipe_state.push_back(std::make_shared<PIPELINE_STATE>(this, &pCreateInfos[i],
                                                                               GetRenderPassShared(pCreateInfos[i].renderPass),
                                                                                GetPipelineLayoutShared(pCreateInfos[i].layout)));
         } else if (enabled_features.dynamic_rendering_features.dynamicRendering) {
              auto dynamic_rendering = LvlFindInChain<VkPipelineRenderingCreateInfoKHR>(pCreateInfos[i].pNext);
              cgpl_state->pipe_state.push_back(std::make_shared<PIPELINE_STATE>(this, &pCreateInfos[i],
                                                                                std::make_shared<RENDER_PASS_STATE>(dynamic_rendering),
                                                                                GetPipelineLayoutShared(pCreateInfos[i].layout)));
          } else {
              cgpl_state->pipe_state.push_back(std::shared_ptr<PIPELINE_STATE>());
          }
      }

```

jeremyg-lunarg replied:
I just pushed this fix.

jeremyg-lunarg reviewed (commented):
It looks like we're down to a few things. I am mainly concerned that the new test cases (including in PR #3525) do not crash:

- VkLayerTest.CreateGraphicsPipelineNullRenderPass
- VkLayerTest.CreateGraphicsPipelineDynamicRenderingNoInfo
- VkPositiveLayerTest.CreateGraphicsPipelineDynamicRendering

I've noted in line where I think the crashes are coming from, but if the tests are wrong lets fix them instead.

There's also an offline discussion about the plan to finish the validation and tests for this extension that you should have received.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 15675.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 15680.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5481 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 5481 passed.

jeremyg-lunarg reviewed (approved):
LGTM.  We'll need to address additional VUIDs and missing test cases in separate PRs. I'll be opening issues to track.
